driver-pulsar/src/main/java/io/nosqlbench/driver/pulsar/ops/PulsarConsumerOp.java

```diff
@@ -11,106 +12,260 @@
+                        String avroDefStr = pulsarSchema.getSchemaInfo().getSchemaDefinition();
+                        org.apache.avro.Schema avroSchema =
+                            AvroUtil.GetSchema_ApacheAvro(avroDefStr);
+                        org.apache.avro.generic.GenericRecord avroGenericRecord =
+                            AvroUtil.GetGenericRecord_ApacheAvro(avroSchema, message.getData());
 
-                    org.apache.avro.generic.GenericRecord avroGenericRecord =
-                        AvroUtil.GetGenericRecord_ApacheAvro(avroSchema, message.getData());
+                        logger.debug("Sync message received: msg-key={}; msg-properties={}; msg-payload={}",
+                            message.getKey(),
+                            message.getProperties(),
+                            avroGenericRecord.toString());
+                    }
+                    else {
+                        logger.debug("Sync message received: msg-key={}; msg-properties={}; msg-payload={}",
+                            message.getKey(),
+                            message.getProperties(),
+                            new String(message.getData()));
+                    }
+                }
 
-                    logger.debug("msg-key={}  msg-payload={}", message.getKey(), avroGenericRecord.toString());
+                // keep track end-to-end message processing latency
+                long e2eMsgLatency = System.currentTimeMillis() - message.getPublishTime();
+                if (e2eMsgProc) {
+                    e2eMsgProcLatencyHistogram.update(e2eMsgLatency);
                 }
-            } else {
-                if (logger.isDebugEnabled()) {
-                    logger.debug("msg-key={}  msg-payload={}", message.getKey(), new String(message.getData()));
+
+                // keep track of message ordering and message loss
+                if (seqTracking) {
+                    String msgSeqIdStr = message.getProperties().get(PulsarActivityUtil.MSG_SEQUENCE_ID);
+                    curMsgSeqId = Long.parseLong(msgSeqIdStr);
+
+                    // normal case: message sequence id is monotonically increasing by 1
+                    if ((curMsgSeqId - prevMsgSeqid) == 1) {
+                        prevMsgSeqid = curMsgSeqId;
+                    }
+                    else {
+                        // abnormal case: out of ordering
+                        if (curMsgSeqId < prevMsgSeqid) {
+                            throw new RuntimeException("Detected message ordering is not guaranteed. Older messages are received earlier!");
+                        }
+                        // abnormal case: message loss
+                        else if ( (curMsgSeqId - prevMsgSeqid) > 1 ) {
+                            throw new RuntimeException("Detected message sequence id gap. Some published messages are not received!");
+                        }
+                    }
```

Comment:
It would be useful to have counters for detected message loss / out-of-order delivery. There is also the case of duplicate message delivery which is expected behavior unless de-duplication feature of Pulsar is used.

duplicates can happen in producing as explained in https://jack-vanlightly.com/blog/2018/10/21/how-to-not-lose-messages-on-an-apache-pulsar-cluster :
> Finally, regarding duplicates. The Pulsar client will automatically retry messages on failure up to the send timeout. This means that duplicates can be created when a message successfully gets written to BookKeeper but the Pulsar broker dies or fails before it can send an ack to the client.

If we'd like to support a way where the checks are made in a streaming fashion, I guess the algorithm would have to support streaming by using a window (or sliding window) for determining whether a message is lost or delivered out of order. 
Duplicate detection probably doesn't require a windowing algorithm, since checking for a previous sequence id would be sufficient for duplicate detection?

@yabinmeng What are your thoughts in improving this area?

Reply:
@lhotari here are my 2 cents. 

1). Regarding counter metrics for message loss / out-of-order delivery, do we really need it? My thinking here is as long as we detect one such occurrence, that's a critical error and we should stop. The criticality of having one such occurrence is as big as have many and there is no point to count how many such occurrences are
2). Regarding duplication, would it be considered an error for the application (if de-duplication is not enabled)? From Pulsar's perspective, this is a normal behavior in order to avoid message loss. I guess what we should add in NB Pulsar driver is if  de-duplication is enabled, then we should catch the case if the consumer receives duplicate messages.
3). Regarding your "streaming" comment with windowing, I don't fully understand and let's chat.